lisa/sut_orchestrator/azure/features.py

```diff
@@ -689,22 +690,51 @@ def setup_rdma(self) -> None:
         if len(devices) > 1:
             # upgrade waagent to latest version to resolve
             # multiple ib devices not getting ip address issue
-            waagent.upgrade_from_source()
+
+            # Upgrade to v2.9.0.4 since the latest v2.9.1.1 version
+            # does not successfully assign IP over the IB interface
+            waagent.upgrade_from_source("v2.9.0.4")
         # Update waagent.conf
         sed = self._node.tools[Sed]
+        rdma_config = "OS.EnableRDMA=y"
+        waagent_config_path = "/etc/waagent.conf"
+
+        # CBLMariner's waagent configuration does not specify OS.EnableRDMA
+        # and thus need to manually append it for RDMA support
+        if isinstance(self._node.os, CBLMariner):
+            # Check whether OS.EnableRDMA=y is already specified
+            cat = self._node.tools[Cat]
+            if rdma_config not in cat.read(waagent_config_path, sudo=True):
+                sed.append(
+                    text=rdma_config,
+                    file=waagent_config_path,
+                    sudo=True,
+                )
+
         sed.substitute(
-            regexp="# OS.EnableRDMA=y",
-            replacement="OS.EnableRDMA=y",
-            file="/etc/waagent.conf",
+            regexp=f"# {rdma_config}",
+            replacement=rdma_config,
+            file=waagent_config_path,
             sudo=True,
         )
         sed.substitute(
             regexp="# AutoUpdate.Enabled=y",
             replacement="AutoUpdate.Enabled=y",
-            file="/etc/waagent.conf",
+            file=waagent_config_path,
             sudo=True,
         )
-        waagent.restart()
+
+        # CBLMariner uses the Mellanox inbox driver instead of the Mellanox OFED driver,
+        # which requires a reboot of the VM for WALinuxAgent to enable RDMA.
+        if isinstance(self._node.os, CBLMariner):
+            self._node.reboot()
```

Comment:
I actually tried just restarting waagent.service without rebooting and that does not work. Suspect that it might be related to using the inbox driver. This [instruction](https://techcommunity.microsoft.com/t5/azure-compute-blog/configuring-infiniband-for-ubuntu-hpc-and-gpu-vms/ba-p/1221351) also says to reboot the VM after editing waagent conf for the inbox driver scenario.

Reply:
I think the comments is unclear, probably the reboot is for loading some modules, not for waagent.

Reply:
After further investigation, seems like the reboot is not required. The actual root-cause is that required kernel modules: ib_uverbs, rdma_ucm, ib_umad and ib_ipoib need to be loaded so that walinuxagent can successfully enable RDMA. Re-factor the implementation to check and load these four kernel modules and then restart the service.